# ***Online News Popularity Project***

##### ***Name*** : _Allen Daniel Raju_
##### ***Batch*** :  _DSML - D53_ 
##### ***Organization*** : _Entri Elevate Coding_
##### ***Date*** : _18-04-2025_

### Modeling

In [1]:
%matplotlib inline

In [2]:
import numpy as np
import scipy as sp
import scipy.stats as stats
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge, Lasso
from sklearn.linear_model import RidgeCV, LassoCV
from sklearn.ensemble import RandomForestRegressor
import statsmodels.formula.api as smf
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [3]:
# Load dataset
df = pd.read_csv('df_cleaned.csv')

#### Simple Linear Regression

In [4]:
df.head()

,n_tokens_title,n_tokens_content,n_non_stop_unique_tokens,num_hrefs,num_self_hrefs,num_imgs,num_videos,average_token_length,num_keywords,kw_min_min,...,avg_negative_polarity,min_negative_polarity,max_negative_polarity,title_subjectivity,title_sentiment_polarity,abs_title_subjectivity,abs_title_sentiment_polarity,shares,data_channel,weekday_is
0,12.0,219.0,0.815385,4.0,2.0,1.0,0.0,4.680365,5.0,0.0,...,-0.350000,-0.600,-0.200000,0.500000,-0.187500,0.000000,0.187500,593,2,1
1,9.0,255.0,0.791946,3.0,1.0,1.0,0.0,4.913725,4.0,0.0,...,-0.118750,-0.125,-0.100000,0.000000,0.000000,0.500000,0.000000,711,3,1
2,9.0,211.0,0.663866,3.0,1.0,1.0,0.0,4.393365,6.0,0.0,...,-0.466667,-0.800,-0.133333,0.000000,0.000000,0.500000,0.000000,1500,3,1
3,9.0,531.0,0.665635,9.0,0.0,1.0,0.0,4.404896,7.0,0.0,...,-0.369697,-0.600,-0.166667,0.000000,0.000000,0.500000,0.000000,1200,2,1
4,13.0,1072.0,0.540890,19.0,19.0,20.0,0.0,4.682836,7.0,0.0,...,-0.220192,-0.500,-0.050000,0.454545,0.136364,0.045455,0.136364,505,5,1


In [5]:
df.columns

Index(['n_tokens_title', 'n_tokens_content', 'n_non_stop_unique_tokens',
       'num_hrefs', 'num_self_hrefs', 'num_imgs', 'num_videos',
       'average_token_length', 'num_keywords', 'kw_min_min', 'kw_max_min',
       'kw_min_max', 'kw_max_max', 'kw_avg_max', 'kw_min_avg', 'kw_max_avg',
       'self_reference_min_shares', 'self_reference_max_shares',
       'self_reference_avg_sharess', 'LDA_00', 'LDA_01', 'LDA_02', 'LDA_03',
       'LDA_04', 'global_subjectivity', 'global_sentiment_polarity',
       'global_rate_positive_words', 'global_rate_negative_words',
       'rate_positive_words', 'rate_negative_words', 'avg_positive_polarity',
       'min_positive_polarity', 'max_positive_polarity',
       'avg_negative_polarity', 'min_negative_polarity',
       'max_negative_polarity', 'title_subjectivity',
       'title_sentiment_polarity', 'abs_title_subjectivity',
       'abs_title_sentiment_polarity', 'shares', 'data_channel', 'weekday_is'],
      dtype='object')

In [6]:
# Predictors in the dataset
predictors = ['n_tokens_title', 'n_tokens_content', 'n_non_stop_unique_tokens',
       'num_hrefs', 'num_self_hrefs', 'num_imgs', 'num_videos',
       'average_token_length', 'num_keywords', 'kw_min_min', 'kw_max_min',
       'kw_min_max', 'kw_max_max', 'kw_avg_max', 'kw_min_avg', 'kw_max_avg',
       'self_reference_min_shares', 'self_reference_max_shares',
       'self_reference_avg_sharess', 'LDA_00', 'LDA_01', 'LDA_02', 'LDA_03',
       'LDA_04', 'global_subjectivity', 'global_sentiment_polarity',
       'global_rate_positive_words', 'global_rate_negative_words',
       'rate_positive_words', 'rate_negative_words', 'avg_positive_polarity',
       'min_positive_polarity', 'max_positive_polarity',
       'avg_negative_polarity', 'min_negative_polarity',
       'max_negative_polarity', 'title_subjectivity',
       'title_sentiment_polarity', 'abs_title_subjectivity',
       'abs_title_sentiment_polarity', 'data_channel', 'weekday_is']

In [7]:
# Extract features (X) and target (y)
X = df[predictors]
y = df['shares']

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [8]:
# Shape of train and test datasets
print(X_train.shape)
print(X_test.shape)

(25756, 42)
(12687, 42)


In [9]:
# Print shape of y objects
print(y_train.shape)
print(y_test.shape)

(25756,)
(12687,)


In [10]:
# Create DataFrame with training dataset
x_df_train = pd.DataFrame(data=X_train, columns=predictors)
x_df_train.head()

,n_tokens_title,n_tokens_content,n_non_stop_unique_tokens,num_hrefs,num_self_hrefs,num_imgs,num_videos,average_token_length,num_keywords,kw_min_min,...,max_positive_polarity,avg_negative_polarity,min_negative_polarity,max_negative_polarity,title_subjectivity,title_sentiment_polarity,abs_title_subjectivity,abs_title_sentiment_polarity,data_channel,weekday_is
35308,12.0,294.0,0.820359,7.0,5.0,2.0,1.0,4.625850,7.0,-1.0,...,1.0,-0.332143,-0.400000,-0.1250,0.00000,0.000000,0.50000,0.000000,2,3
2595,11.0,284.0,0.762500,1.0,0.0,1.0,1.0,4.690141,5.0,217.0,...,0.8,-0.194444,-0.300000,-0.1000,0.60000,0.200000,0.10000,0.200000,3,4
6100,10.0,254.0,0.745455,7.0,5.0,1.0,0.0,4.748031,6.0,4.0,...,0.7,-0.133333,-0.166667,-0.1000,0.00000,0.000000,0.50000,0.000000,3,2
11827,9.0,292.0,0.750000,6.0,5.0,0.0,17.0,4.602740,5.0,4.0,...,1.0,-0.475000,-0.700000,-0.2000,0.49513,0.211039,0.00487,0.211039,4,4
12758,11.0,306.0,0.741573,15.0,2.0,15.0,0.0,4.205882,10.0,4.0,...,1.0,-0.473438,-1.000000,-0.1875,0.00000,0.000000,0.50000,0.000000,7,3


In [11]:
# Training series with shares
y_df_train = pd.DataFrame(data=y_train, columns=['shares']) 
y_df_train.head()

,shares
35308,1400
2595,2300
6100,968
11827,1500
12758,74100


In [12]:
# Concatenate x and y together to create dataframe of training set
train_df = pd.concat([x_df_train, y_df_train], axis=1, sort=False)
train_df.head()

,n_tokens_title,n_tokens_content,n_non_stop_unique_tokens,num_hrefs,num_self_hrefs,num_imgs,num_videos,average_token_length,num_keywords,kw_min_min,...,avg_negative_polarity,min_negative_polarity,max_negative_polarity,title_subjectivity,title_sentiment_polarity,abs_title_subjectivity,abs_title_sentiment_polarity,data_channel,weekday_is,shares
35308,12.0,294.0,0.820359,7.0,5.0,2.0,1.0,4.625850,7.0,-1.0,...,-0.332143,-0.400000,-0.1250,0.00000,0.000000,0.50000,0.000000,2,3,1400
2595,11.0,284.0,0.762500,1.0,0.0,1.0,1.0,4.690141,5.0,217.0,...,-0.194444,-0.300000,-0.1000,0.60000,0.200000,0.10000,0.200000,3,4,2300
6100,10.0,254.0,0.745455,7.0,5.0,1.0,0.0,4.748031,6.0,4.0,...,-0.133333,-0.166667,-0.1000,0.00000,0.000000,0.50000,0.000000,3,2,968
11827,9.0,292.0,0.750000,6.0,5.0,0.0,17.0,4.602740,5.0,4.0,...,-0.475000,-0.700000,-0.2000,0.49513,0.211039,0.00487,0.211039,4,4,1500
12758,11.0,306.0,0.741573,15.0,2.0,15.0,0.0,4.205882,10.0,4.0,...,-0.473438,-1.000000,-0.1875,0.00000,0.000000,0.50000,0.000000,7,3,74100


In [13]:
# Create DataFrame with test dataset
x_df_test = pd.DataFrame(data=X_test, columns=predictors)
x_df_test.head()

,n_tokens_title,n_tokens_content,n_non_stop_unique_tokens,num_hrefs,num_self_hrefs,num_imgs,num_videos,average_token_length,num_keywords,kw_min_min,...,max_positive_polarity,avg_negative_polarity,min_negative_polarity,max_negative_polarity,title_subjectivity,title_sentiment_polarity,abs_title_subjectivity,abs_title_sentiment_polarity,data_channel,weekday_is
5532,9.0,218.0,0.843750,2.0,1.0,0.0,1.0,4.353211,7.0,4.0,...,0.6,-0.409524,-0.666667,-0.071429,0.0,0.0000,0.5,0.0000,2,4
15242,11.0,268.0,0.835616,5.0,3.0,1.0,0.0,4.492537,4.0,4.0,...,0.6,-0.118750,-0.187500,-0.050000,0.1,0.3000,0.4,0.3000,6,3
35329,14.0,851.0,0.633858,15.0,6.0,1.0,0.0,4.584019,7.0,-1.0,...,0.8,-0.307568,-0.800000,-0.050000,0.7,0.1125,0.2,0.1125,5,3
9288,9.0,725.0,0.536748,19.0,6.0,1.0,0.0,4.590345,10.0,4.0,...,1.0,-0.246875,-0.500000,-0.100000,0.9,-0.6000,0.4,0.6000,6,3
14634,8.0,616.0,0.753846,10.0,0.0,1.0,0.0,4.683442,7.0,4.0,...,1.0,-0.337976,-0.625000,-0.071429,0.0,0.0000,0.5,0.0000,2,5


In [14]:
# Test series with shares
y_df_test = pd.DataFrame(data=y_test, columns=['shares']) 
y_df_test.head()

,shares
5532,1500
15242,1100
35329,1600
9288,9500
14634,682


In [15]:
# Concatenate x and y together to create dataframe of test set
test_df = pd.concat([x_df_test, y_df_test], axis=1, sort=False)
test_df.head()

,n_tokens_title,n_tokens_content,n_non_stop_unique_tokens,num_hrefs,num_self_hrefs,num_imgs,num_videos,average_token_length,num_keywords,kw_min_min,...,avg_negative_polarity,min_negative_polarity,max_negative_polarity,title_subjectivity,title_sentiment_polarity,abs_title_subjectivity,abs_title_sentiment_polarity,data_channel,weekday_is,shares
5532,9.0,218.0,0.843750,2.0,1.0,0.0,1.0,4.353211,7.0,4.0,...,-0.409524,-0.666667,-0.071429,0.0,0.0000,0.5,0.0000,2,4,1500
15242,11.0,268.0,0.835616,5.0,3.0,1.0,0.0,4.492537,4.0,4.0,...,-0.118750,-0.187500,-0.050000,0.1,0.3000,0.4,0.3000,6,3,1100
35329,14.0,851.0,0.633858,15.0,6.0,1.0,0.0,4.584019,7.0,-1.0,...,-0.307568,-0.800000,-0.050000,0.7,0.1125,0.2,0.1125,5,3,1600
9288,9.0,725.0,0.536748,19.0,6.0,1.0,0.0,4.590345,10.0,4.0,...,-0.246875,-0.500000,-0.100000,0.9,-0.6000,0.4,0.6000,6,3,9500
14634,8.0,616.0,0.753846,10.0,0.0,1.0,0.0,4.683442,7.0,4.0,...,-0.337976,-0.625000,-0.071429,0.0,0.0000,0.5,0.0000,2,5,682


In [16]:
# Use Statsmodel to create simple linear regression model of shares based on a single predictor 
# train model on the training set 

# Loop through predictors and fit linear models
for i in predictors:
    lin_model = smf.ols(formula=f'shares ~ {i}', data=train_df).fit()
    print(f"Dependent variable: {i}")
    print(f"R-squared: {lin_model.rsquared}")
    print(f"Coefficient 0 (Intercept): {lin_model.params[0]}")
    print(f"Coefficient 1 ({i}): {lin_model.params[1]}")
    print()

Dependent variable: n_tokens_title
R-squared: 0.0001515237439502215
Coefficient 0 (Intercept): 2763.7066860373816
Coefficient 1 (n_tokens_title): 41.38297036208207

Dependent variable: n_tokens_content
R-squared: 1.616089151390554e-07
Coefficient 0 (Intercept): 3189.8210198786023
Coefficient 1 (n_tokens_content): 0.006101657911188135

Dependent variable: n_non_stop_unique_tokens
R-squared: 4.6289012812783525e-06
Coefficient 0 (Intercept): 3190.562721499027
Coefficient 1 (n_non_stop_unique_tokens): 3.754671787264772

Dependent variable: num_hrefs
R-squared: 0.006000091722709655
Coefficient 0 (Intercept): 2651.9294168788856
Coefficient 1 (num_hrefs): 48.15881586146109

Dependent variable: num_self_hrefs
R-squared: 8.405814721945681e-06
Coefficient 0 (Intercept): 3175.4156501004063
Coefficient 1 (num_self_hrefs): 5.250899905331114

Dependent variable: num_imgs
R-squared: 0.004309108074239676
Coefficient 0 (Intercept): 2937.6890883089864
Coefficient 1 (num_imgs): 56.208635118981945

Depend

C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 0 (Intercept): {lin_model.params[0]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 1 ({i}): {lin_model.params[1]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `s

Dependent variable: kw_max_max
R-squared: 0.0001585300416966362
Coefficient 0 (Intercept): 2883.638381215238
Coefficient 1 (kw_max_max): 0.00041209459525493183

Dependent variable: kw_avg_max
R-squared: 0.002761497059635354
Coefficient 0 (Intercept): 2473.559085330373
Coefficient 1 (kw_avg_max): 0.002822519902888512

Dependent variable: kw_min_avg
R-squared: 0.0022968557146618718
Coefficient 0 (Intercept): 2861.6449487364825
Coefficient 1 (kw_min_avg): 0.3007999101907428

Dependent variable: kw_max_avg
R-squared: 0.008831122613249698
Coefficient 0 (Intercept): 2543.359809775412
Coefficient 1 (kw_max_avg): 0.11638521701139698

Dependent variable: self_reference_min_shares
R-squared: 0.0038962366980352936
Coefficient 0 (Intercept): 3100.1047212012836
Coefficient 1 (self_reference_min_shares): 0.022840463860789178

Dependent variable: self_reference_max_shares
R-squared: 0.0033943264041189636
Coefficient 0 (Intercept): 3089.086281614372
Coefficient 1 (self_reference_max_shares): 0.0096286

C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 0 (Intercept): {lin_model.params[0]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 1 ({i}): {lin_model.params[1]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `s

Dependent variable: global_subjectivity
R-squared: 0.009733678847423155
Coefficient 0 (Intercept): -430.4446250680072
Coefficient 1 (global_subjectivity): 7936.358317846327

Dependent variable: global_sentiment_polarity
R-squared: 0.00022116305101993472
Coefficient 0 (Intercept): 3057.9778371680495
Coefficient 1 (global_sentiment_polarity): 1097.6511014259922

Dependent variable: global_rate_positive_words
R-squared: 0.00023226595455005938
Coefficient 0 (Intercept): 2920.383741535202
Coefficient 1 (global_rate_positive_words): 6681.068935942345

Dependent variable: global_rate_negative_words
R-squared: 0.0005961799695456405
Coefficient 0 (Intercept): 2913.3725985746023
Coefficient 1 (global_rate_negative_words): 16352.290089652484

Dependent variable: rate_positive_words
R-squared: 8.418653993780278e-05
Coefficient 0 (Intercept): 3498.0439720167215
Coefficient 1 (rate_positive_words): -433.3180358208961

Dependent variable: rate_negative_words
R-squared: 8.583326579880701e-05
Coefficie

C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 0 (Intercept): {lin_model.params[0]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 1 ({i}): {lin_model.params[1]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `s

Dependent variable: min_negative_polarity
R-squared: 0.0017330748687242048
Coefficient 0 (Intercept): 2628.71950034944
Coefficient 1 (min_negative_polarity): -1050.0873043785075

Dependent variable: max_negative_polarity
R-squared: 0.0007739818092717954
Coefficient 0 (Intercept): 2964.064299733359
Coefficient 1 (max_negative_polarity): -2073.7589360703087

Dependent variable: title_subjectivity
R-squared: 0.0019803986610165136
Coefficient 0 (Intercept): 2920.647900607458
Coefficient 1 (title_subjectivity): 971.3835981890277

Dependent variable: title_sentiment_polarity
R-squared: 0.0012705721495491762
Coefficient 0 (Intercept): 3124.4450398637923
Coefficient 1 (title_sentiment_polarity): 951.708313995513

Dependent variable: abs_title_subjectivity
R-squared: 7.42457573199573e-08
Coefficient 0 (Intercept): 3196.747406745923
Coefficient 1 (abs_title_subjectivity): -10.190140616256855

Dependent variable: abs_title_sentiment_polarity
R-squared: 0.0026217984032116393
Coefficient 0 (Interce

C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 0 (Intercept): {lin_model.params[0]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"Coefficient 1 ({i}): {lin_model.params[1]}")
C:\Users\mktgs\AppData\Local\Temp\ipykernel_12524\2706907561.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `s

In [17]:
LDA_03_model = smf.ols(formula=f'shares~ LDA_03', data=train_df).fit()
sm.qqplot(LDA_03_model.resid, line='s');

#### Multilinear Regression

In [18]:
# Turn list of predictors into string with plus signs that can be entered as model formula
def list_to_string(predictor_array):
    listToStr = '+'.join(map(str, predictor_array))
    print(listToStr)
    return listToStr

In [19]:
# Create a multilinear model that includes all predictors 
# Find coefficients that reduce the sum of least squares 
ml_model = smf.ols(formula='shares ~ {}'.format(list_to_string(predictors)), data=train_df).fit()
print(ml_model.summary())

n_tokens_title+n_tokens_content+n_non_stop_unique_tokens+num_hrefs+num_self_hrefs+num_imgs+num_videos+average_token_length+num_keywords+kw_min_min+kw_max_min+kw_min_max+kw_max_max+kw_avg_max+kw_min_avg+kw_max_avg+self_reference_min_shares+self_reference_max_shares+self_reference_avg_sharess+LDA_00+LDA_01+LDA_02+LDA_03+LDA_04+global_subjectivity+global_sentiment_polarity+global_rate_positive_words+global_rate_negative_words+rate_positive_words+rate_negative_words+avg_positive_polarity+min_positive_polarity+max_positive_polarity+avg_negative_polarity+min_negative_polarity+max_negative_polarity+title_subjectivity+title_sentiment_polarity+abs_title_subjectivity+abs_title_sentiment_polarity+data_channel+weekday_is
                            OLS Regression Results                            
Dep. Variable:                 shares   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.037
Method:                 Least Squares   F-st

In [20]:
# Create a QQ plot for the residuals
sm.qqplot(ml_model.resid, line='s');

#### Enhancing Predictive Accuracy Through Regularization: Exploring Ridge and Lasso Regression

In [21]:
# Instantiate Ridge and Lasso regression models
# Alpha is a tuning parameter that determines strength of the regularization applied to the model
ridge_model = Ridge(alpha=1.0) 
lasso_model = Lasso(alpha=1.0)

# Fit the models to the training data
ridge_model.fit(X_train, y_train)
lasso_model.fit(X_train, y_train)

# Evaluate the models on the test data
ridge_r2 = ridge_model.score(X_test, y_test)
lasso_r2 = lasso_model.score(X_test, y_test)

print(f"R-squared for Ridge regression: {ridge_r2:.4f}")
print(f"R-squared for Lasso regression: {lasso_r2:.4f}")

R-squared for Ridge regression: 0.0281
R-squared for Lasso regression: 0.0284


#### Ridge and Lasso regression models using cross-validation for alpha selection

In [22]:
# Create Ridge and Lasso models with a range of alpha values
ridge_model = RidgeCV(alphas=[0.01, 0.1, 1.0, 5.0, 10.0, 20.0, 50.0])
lasso_model = LassoCV(alphas=[0.01, 0.1, 1.0, 5.0, 10.0, 20.0, 50.0], max_iter=10000)

# Fit the models to the training data
ridge_model.fit(X_train, y_train)
lasso_model.fit(X_train, y_train)

# Get the best alpha values from the models determined from cross-validation process
best_ridge_alpha = ridge_model.alpha_
best_lasso_alpha = lasso_model.alpha_

print(f"Best alpha for Ridge: {best_ridge_alpha:.4f}")
print(f"Best alpha for Lasso: {best_lasso_alpha:.4f}")

Best alpha for Ridge: 50.0000
Best alpha for Lasso: 10.0000


Now that I have the best alpha values, I will update the Ridge and Lasso models with this value and assess its performance on the test data. This iterative approach will allow me to strike the right balance between regularization and model fit, ultimately leading to a more accurate predictive model.

In [23]:
# Instantiate Ridge and Lasso regression models
ridge_model = Ridge(alpha=best_ridge_alpha) 
lasso_model = Lasso(alpha=best_lasso_alpha)

# Fit the models to the training data
ridge_model.fit(X_train, y_train)
lasso_model.fit(X_train, y_train)

# Evaluate the models on the test data
ridge_r2 = ridge_model.score(X_test, y_test)
lasso_r2 = lasso_model.score(X_test, y_test)

print(f"R-squared for Ridge regression: {ridge_r2:.4f}")
print(f"R-squared for Lasso regression: {lasso_r2:.4f}")

R-squared for Ridge regression: 0.0284
R-squared for Lasso regression: 0.0291


#### Multilinear Regression with Statistically Significant Predictors

In [24]:
significant_predictors = ['n_tokens_title', 'n_non_stop_unique_tokens', 'num_hrefs',
                          'num_self_hrefs', 'num_imgs', 'average_token_length', 'num_keywords',
                          'kw_max_min', 'kw_min_max', 'kw_avg_max', 'kw_min_avg', 'kw_max_avg',
                          'self_reference_min_shares', 'self_reference_max_shares', 'LDA_00',
                          'LDA_01', 'LDA_02', 'LDA_03', 'LDA_04', 'global_subjectivity', 
                          'title_sentiment_polarity', 'abs_title_subjectivity', 'data_channel'
                         ]

In [25]:
# Create a DataFrame with only significant predictors
X_train_significant = X_train[significant_predictors]
X_train_significant.insert(0, 'shares', y_train)  # Insert the target variable as the first column

# Initialize the list to keep track of selected predictors
selected_predictors = []

# Initialize the formula
formula = "shares ~ 1"  # Starting with intercept only

# Initialize variables to keep track of the best R-squared and predictor
best_r_squared = 0.0
best_predictor = None

# Perform forward stepwise regression
for _ in range(len(significant_predictors)):
    remaining_predictors = list(set(significant_predictors) - set(selected_predictors))
    r_squared_values = []
    
    for predictor in remaining_predictors:
        predictors = selected_predictors + [predictor]
        temp_formula = formula + " + " + " + ".join(predictors)
        
        temp_model = smf.ols(formula=temp_formula, data=X_train_significant).fit()
        r_squared = temp_model.rsquared_adj
        r_squared_values.append((predictor, r_squared))

    # Find the predictor with the highest increase in R-squared
    best_predictor, max_r_squared_increase = max(r_squared_values, key=lambda x: x[1])
    
    # Add the best predictor to the model
    selected_predictors.append(best_predictor)
    
    # Update the formula
    formula = formula + " + " + best_predictor

    # Update the best R-squared value
    best_r_squared = max_r_squared_increase
    
    # Print the current step's information
    print(f"Step {_+1}: Added predictor '{best_predictor}' (Adjusted R-squared: {best_r_squared:.4f})")

# Fit the final model
final_model = smf.ols(formula=formula, data=X_train_significant).fit()

print(final_model.summary())

Step 1: Added predictor 'LDA_03' (Adjusted R-squared: 0.0136)
Step 2: Added predictor 'kw_max_avg' (Adjusted R-squared: 0.0186)
Step 3: Added predictor 'global_subjectivity' (Adjusted R-squared: 0.0228)
Step 4: Added predictor 'self_reference_min_shares' (Adjusted R-squared: 0.0254)
Step 5: Added predictor 'num_hrefs' (Adjusted R-squared: 0.0281)
Step 6: Added predictor 'LDA_02' (Adjusted R-squared: 0.0294)
Step 7: Added predictor 'data_channel' (Adjusted R-squared: 0.0318)
Step 8: Added predictor 'title_sentiment_polarity' (Adjusted R-squared: 0.0323)
Step 9: Added predictor 'num_self_hrefs' (Adjusted R-squared: 0.0329)
Step 10: Added predictor 'num_imgs' (Adjusted R-squared: 0.0332)
Step 11: Added predictor 'n_tokens_title' (Adjusted R-squared: 0.0336)
Step 12: Added predictor 'self_reference_max_shares' (Adjusted R-squared: 0.0339)
Step 13: Added predictor 'average_token_length' (Adjusted R-squared: 0.0341)
Step 14: Added predictor 'num_keywords' (Adjusted R-squared: 0.0343)
Step 15

In [26]:
# Create a QQ plot for the residuals
sm.qqplot(final_model.resid, line='s');

In [27]:
significant_predictors2 = ['kw_max_avg', 'global_subjectivity', 'self_reference_min_shares',
                          'num_hrefs', 'data_channel', 'title_sentiment_polarity', 'num_self_hrefs',
                          'num_imgs', 'n_tokens_title', 'average_token_length', 'num_keywords',
                          'kw_min_avg'
                         ]

In [28]:
# Create a DataFrame with only significant predictors
X_train_significant = X_train[significant_predictors2]
X_train_significant.insert(0, 'shares', y_train)  # Insert the target variable as the first column

# Initialize the list to keep track of selected predictors
selected_predictors = []

# Initialize the formula
formula = "shares ~ 1"  # Starting with intercept only

# Initialize variables to keep track of the best R-squared and predictor
best_r_squared = 0.0
best_predictor = None

# Perform forward stepwise regression
for _ in range(len(significant_predictors2)):
    remaining_predictors = list(set(significant_predictors2) - set(selected_predictors))
    r_squared_values = []
    
    for predictor in remaining_predictors:
        predictors = selected_predictors + [predictor]
        temp_formula = formula + " + " + " + ".join(predictors)
        
        temp_model = smf.ols(formula=temp_formula, data=X_train_significant).fit()
        r_squared = temp_model.rsquared_adj
        r_squared_values.append((predictor, r_squared))

    # Find the predictor with the highest increase in R-squared
    best_predictor, max_r_squared_increase = max(r_squared_values, key=lambda x: x[1])
    
    # Add the best predictor to the model
    selected_predictors.append(best_predictor)
    
    # Update the formula
    formula = formula + " + " + best_predictor

    # Update the best R-squared value
    best_r_squared = max_r_squared_increase
    
    # Print the current step's information
    print(f"Step {_+1}: Added predictor '{best_predictor}' (Adjusted R-squared: {best_r_squared:.4f})")

# Fit the final model
final_model2 = smf.ols(formula=formula, data=X_train_significant).fit()

print(final_model2.summary())
print(formula)

Step 1: Added predictor 'global_subjectivity' (Adjusted R-squared: 0.0097)
Step 2: Added predictor 'kw_max_avg' (Adjusted R-squared: 0.0166)
Step 3: Added predictor 'num_hrefs' (Adjusted R-squared: 0.0202)
Step 4: Added predictor 'self_reference_min_shares' (Adjusted R-squared: 0.0231)
Step 5: Added predictor 'data_channel' (Adjusted R-squared: 0.0244)
Step 6: Added predictor 'num_imgs' (Adjusted R-squared: 0.0254)
Step 7: Added predictor 'average_token_length' (Adjusted R-squared: 0.0261)
Step 8: Added predictor 'num_self_hrefs' (Adjusted R-squared: 0.0269)
Step 9: Added predictor 'title_sentiment_polarity' (Adjusted R-squared: 0.0276)
Step 10: Added predictor 'kw_min_avg' (Adjusted R-squared: 0.0282)
Step 11: Added predictor 'num_keywords' (Adjusted R-squared: 0.0286)
Step 12: Added predictor 'n_tokens_title' (Adjusted R-squared: 0.0289)
                            OLS Regression Results                            
Dep. Variable:                 shares   R-squared:                   

In [29]:
# Create a QQ plot for the residuals
sm.qqplot(final_model2.resid, line='s');

#### Exploring Predictor Interactions with Random Forest

In [ ]:
# Initialize and configure the model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

# Fit the model on the training data
rf_model.fit(X_train, y_train)

# Predict on the test data
y_pred_rf = rf_model.predict(X_test)

# Calculate the R-squared value of the model
rf_r2 = rf_model.score(X_test, y_test)

# Get feature importances from the model
feature_importances = rf_model.feature_importances_

# Pair feature names with feature importances
feature_importance_pairs = list(zip(predictors, feature_importances))

# Sort feature importances in descending order
feature_importance_pairs.sort(key=lambda x: x[1], reverse=True)

print("Feature Importances:")
for feature, importance in feature_importance_pairs:
    print(f"{feature}: {importance:.4f}")

print(f"Random Forest R-squared: {rf_r2:.4f}")

To further explore the potential interactions among predictors identified by the Random Forest analysis, I conducted a multilinear regression analysis. The goal was to assess how the inclusion of interaction terms could enhance the model's explanatory power.

In [ ]:
selected_features = [feature for feature, importance in feature_importance_pairs[:12]]

In [ ]:
# Create a new DataFrame with only the selected features for training
train_df_selected = train_df[selected_features + ['shares']]

In [ ]:
# Create interaction terms
interaction_terms = []
for predictor1 in selected_features:
    for predictor2 in selected_features:
        if predictor1 != predictor2:
            interaction_name = f"{predictor1}_{predictor2}"
            train_df_selected[interaction_name] = train_df_selected[predictor1] * train_df_selected[predictor2]
            interaction_terms.append(interaction_name)

# Prepare data with interaction terms
data_with_interactions = pd.concat([train_df_selected, train_df_selected[interaction_terms]], axis=1).copy()

# Run multilinear regression with interaction terms
results = []
for interaction in interaction_terms:
    formula = f'shares ~ {" + ".join(selected_features + [interaction])}'
    ml_model = smf.ols(formula=formula, data=data_with_interactions).fit()
    results.append((interaction, ml_model.rsquared_adj))

results.sort(key=lambda x: x[1], reverse=True)
for interaction_name, adj_r2 in results:
    print(f"Interaction: {interaction_name}, Adjusted R-squared: {adj_r2:.4f}")

In [ ]:
# Top 3 interaction terms
top_interactions = [
    "self_reference_min_shares_global_subjectivity",
    "kw_max_avg_num_imgs",
    "self_reference_min_shares_num_imgs"
]

# Prepare data with top interaction terms
data_with_top_interactions = pd.concat([train_df_selected, train_df_selected[top_interactions]], axis=1)

# Run multilinear regression with top interaction terms
for interaction in top_interactions:
    formula = f'shares ~ {" + ".join(selected_features + [interaction])}'
    ml_model = smf.ols(formula=formula, data=data_with_top_interactions).fit()
    print(f"Interaction: {interaction}")
    print(ml_model.summary())
    # Create a QQ plot for the residuals
    sm.qqplot(ml_model.resid, line='s')

#### Summary of Results & Analysis: Exploring Predictor Interactions with Random Forest

I incorporated interaction terms into the model based on the top-performing interactions from the Random Forest analysis. The selected interaction terms were:

* self_reference_min_shares_global_subjectivity
* kw_max_avg_num_imgs
* self_reference_min_shares_num_imgs

These interactions were chosen due to their relatively higher adjusted R-squared values and their potential to reveal intricate relationships between predictors.

The multilinear regression models with interaction terms exhibited limited improvement in explaining the variance in the target variable, "shares." The adjusted R-squared values for these models was 0.031. While statistically significant, these values indicate that only about 3.1% of the variation in the target variable could be explained by the included predictors and their interactions.

This modest increase in adjusted R-squared as compared to the multilinear model with statistically significant predictors suggests that while these interaction terms contribute to the model's predictive capacity, they do not strongly enhance its overall performance. It's important to note that the inclusion of interaction terms can introduce multicollinearity and increase model complexity, potentially leading to overfitting. Thus far, the model with the highest R-squared is the multilinear model with all predictors.

#### Multicollinearity Detection: VIF

In [ ]:
# Get variables for which to compute VIF and add intercept term
# set all values in intercept term to 1
x_df_train['Intercept'] = 1
x_df_train.head()

In [ ]:
# Compute and view VIF
vif = pd.DataFrame()
# using variance inflation factor from Statsmodel
vif['variables'] = x_df_train.columns
vif["VIF"] = [variance_inflation_factor(x_df_train.values, i) for i in range(x_df_train.shape[1])]
print(vif)

In [ ]:
# drop predictors with VIF greater than 5
columns_to_drop = ['self_reference_min_shares', 'self_reference_max_shares',
                'LDA_00', 'LDA_02', 'LDA_03', 'LDA_04', 'global_sentiment_polarity',
                'global_rate_negative_words', 'avg_negative_polarity']

x_df_train = x_df_train.drop(columns_to_drop, axis=1)
x_df_train.head()

In [ ]:
# Compute and view VIF to ensure all interactions are below 5
vif2 = pd.DataFrame()
# using variance inflation factor from Statsmodel
vif2['variables'] = x_df_train.columns
vif2["VIF"] = [variance_inflation_factor(x_df_train.values, i) for i in range(x_df_train.shape[1])]
print(vif2)

The rate_positive_words and rate_negative_words have a threshold over 5 and will be dropped from the dataset.

In [ ]:
# drop predictors with VIF greater than 5
columns_to_drop = ['rate_positive_words', 'rate_negative_words']

x_df_train = x_df_train.drop(columns_to_drop, axis=1)
x_df_train.head()

In [ ]:
# Compute and view VIF to ensure all interactions are below 5
vif3 = pd.DataFrame()
# using variance inflation factor from Statsmodel
vif3['variables'] = x_df_train.columns
vif3["VIF"] = [variance_inflation_factor(x_df_train.values, i) for i in range(x_df_train.shape[1])]
print(vif3)

Now that VIF for all intereactions is below 5. I can rerun the multilinear model.

In [ ]:
predictors_wo_multicol = [
    'n_tokens_title', 'n_tokens_content', 'n_non_stop_unique_tokens',
    'num_hrefs', 'num_self_hrefs', 'num_imgs',
    'num_videos', 'average_token_length', 'num_keywords',
    'kw_min_min', 'kw_max_min', 'kw_min_max',
    'kw_max_max', 'kw_avg_max', 'kw_min_avg',
    'kw_max_avg', 'self_reference_avg_sharess', 'LDA_01',
    'global_subjectivity', 'global_rate_positive_words', 'avg_positive_polarity',
    'min_positive_polarity', 'max_positive_polarity', 'min_negative_polarity',
    'max_negative_polarity', 'title_subjectivity', 'title_sentiment_polarity',
    'abs_title_subjectivity', 'abs_title_sentiment_polarity', 'data_channel',
    'weekday_is'
]

In [ ]:
no_multicol_model = smf.ols(formula='shares ~ {}'.format(list_to_string(predictors_wo_multicol)), data=train_df).fit()
no_multicol_model.summary()

In [ ]:
# Create a QQ plot for the residuals
sm.qqplot(no_multicol_model.resid, line='s');

#### Polynomial Regression

In [ ]:
def polynomial_regression_summary(data, predictors, target, degree):
    # Construct the polynomial formula
    poly_formula = f'{target} ~ ' + ' + '.join([f'{col} + np.power({col}, {degree})' for col in predictors])
    
    # Fit the polynomial regression model
    poly_est = smf.ols(poly_formula, data).fit()
    
    # Return the fitted model
    return poly_est

#### Initial Polynomial Regression with All Predictors

In [ ]:
degree = 2
polynomial_regression_summary(train_df, predictors, 'shares', degree).summary()

In [ ]:
degree = 3
polynomial_regression_summary(train_df, predictors, 'shares', degree).summary()

In [ ]:
degree = 4
polynomial_regression_summary(train_df, predictors, 'shares', degree).summary()

In [ ]:
degree = 5
polynomial_regression_summary(train_df, predictors, 'shares', degree).summary()

#### Summary of Results & Analysis: Initial Polynomial Regression

The initial polynomial regression was performed using all available predictor variables. Degrees ranging from 2 to 5 were evaluated to find the best-fitting model. The highest achieved R-squared value was 0.043, which corresponded to a polynomial degree of 2. This modest improvement over the best R-squared value obtained from multilinear regression (0.038) suggested that polynomial terms helped capture some of the nonlinear trends in the data

#### Polynomial Regression without Multicollinearity

In [ ]:
degree = 2
polynomial_regression_summary(train_df, predictors_wo_multicol, 'shares', degree).summary()

In [ ]:
degree = 3
polynomial_regression_summary(train_df, predictors_wo_multicol, 'shares', degree).summary()

In [ ]:
degree = 4
polynomial_regression_summary(train_df, predictors_wo_multicol, 'shares', degree).summary()

In [ ]:
degree = 5
polynomial_regression_summary(train_df, predictors_wo_multicol, 'shares', degree).summary()

#### Summary of Results & Analysis: Polynomial Regression without Multicollinearity

Considering the potential impact of multicollinearity on the model's stability and interpretability, polynomial regression was further conducted using predictor variables devoid of significant multicollinearity. This refinement yielded a slightly improved R-squared value of 0.047, again at a polynomial degree of 2. This is the best model so far overall.

#### Polynomial Regression with Focus on Significant Predictors

In [ ]:
degree = 2
polynomial_regression_summary(train_df, significant_predictors, 'shares', degree).summary()

In [ ]:
degree = 3
polynomial_regression_summary(train_df, significant_predictors, 'shares', degree).summary()

In [ ]:
degree = 4
polynomial_regression_summary(train_df, significant_predictors, 'shares', degree).summary()

In [ ]:
degree = 5
polynomial_regression_summary(train_df, significant_predictors, 'shares', degree).summary()

#### Run polynomial regression on 2nd iteration of statistically significant predictors

In [ ]:
degree = 2
polynomial_regression_summary(train_df, significant_predictors2, 'shares', degree).summary()

In [ ]:
degree = 3
polynomial_regression_summary(train_df, significant_predictors2, 'shares', degree).summary()

In [ ]:
degree = 4
polynomial_regression_summary(train_df, significant_predictors2, 'shares', degree).summary()

In [ ]:
degree = 5
polynomial_regression_summary(train_df, significant_predictors2, 'shares', degree).summary()

#### Summary of Results & Analysis: Focus on Significant Predictors

Given the goal of identifying the most influential factors affecting article shares, polynomial regression was employed on the subset of significant predictor variables. Remarkably, this approach produced the highest R-squared value of the entire analysis: 0.050 at a polynomial degree of 2. This underscores the importance of focusing on key predictors to build a more concise and accurate model, although, this model is still not very accurate as the R-squared is closer to 0 than 1.

#### Summary of Results & Analysis: Polynomial Regression Overall

The application of polynomial regression provided valuable insights into the underlying relationships between predictor variables and the response variable. By flexibly accommodating nonlinear trends, this technique improved the model's predictive capability compared to the linear and multilinear regression approaches. The achieved R-squared values highlight the impact of careful predictor selection and the importance of accounting for nonlinearity in understanding the determinants of article shares. The highest R-squared achieved in this entire analysis was 0.05, which does mean that the model accuracy is not very good, as the model is significantly underfit.

### Discussion & Conclusion
In this study, my goal was to explore article sharing behavior in the digital landscape. I went into this project with the hope that I could construct a model that could be used to accurately predict the number of shares an article was likely to get. I explored every technique learned in class and some techniques that I taught myself. From the straightforward linear regression to the more sophisticated multilinear and polynomial regressions, I left no stone unturned in my quest for insights. Although I was not able to construct an accurate model, I did learn several new techniques that I hope to employ in future projects.


#### Learning and Takeaways
In the process of conducting this analysis, several valuable insights have been gleaned. The application of various regression techniques, including linear, multilinear, and polynomial regression, shed light on the complex relationship between predictor variables and article shares.

One key takeaway is the importance of careful predictor selection. Through rigorous analysis, it became evident that certain predictors have a significant impact on article shares. This emphasizes the need for a thoughtful approach to feature engineering, which ultimately enhances model performance.

Furthermore, the implementation of polynomial regression proved to be a crucial step in capturing nonlinear trends. This technique provided the flexibility to model intricate relationships that would have been missed by simpler linear models. By accommodating these nonlinearities, the model was able to make more accurate predictions and achieve a better fit to the data, although, it was ultimately underfit.